### DISCUSSION TITLE
[Script] Don't include `next/script` code into the client bundle if it isn't being used

### DISCUSSION THREAD

**Discussion by housseindjirdeh** [vercel/next.js#64635] [Ideas]
### Describe the feature you'd like to request

`next/script` code shouldn't be imported client-side if the Script component isn't being used ([this](https://github.com/vercel/next.js/blob/1d1ffc88b7529441ffda5341e8303a7afbb1357e/packages/next/client/index.tsx#L256-L259) will likely have to modified)

### Describe the solution you'd like

N/A

### Describe alternatives you've considered

N/A

**Comment by SukkaW** [comment]
The issue has already been mentioned in https://github.com/vercel/next.js/issues/36702

And here is my thought:

- `if (initialData.scriptLoader.length) { const { initScriptLoader } = require('./script') }`, won't help and the `initScriptLoader` will always be included in the bundle, because Webpack can't static analyze `initialData.scriptLoader`.
- An internal environment variable (maybe something like `process.env.__NEXT_SCRIPT_BEING_USED`?) might be needed, and Next.js will have to statically analyze if `next/script` is being used. This might be implemented as a plugin for `next-swc`?

**Comment by SukkaW** [comment]
Another thought: we could read every `[mc]?[jt]sx?` files and check `!includes('next/script') && !includes('next/dist/compiled/client/script')`, and set a corresponding `process.env.__NEXT_SCRIPT_BEING_USED` environment variable.

It is a *really* dirty hack, but it works.

**Comment by housseindjirdeh** [comment]
I agree an environment variable might be the easiest way to ensure the code doesn't get bundled when it should, but I'm still not sure where the best place would be to set that var. 🤔 

~Currently exploring if it would be possible to do it in a webpack plugin, but maybe this is overkill?~ Nvm, `DefinePlugin` can't be used to define env variables after bundling, only before ([details](https://github.com/webpack/webpack/discussions/15071#discussioncomment-1887341))
